### PR TITLE
Reduce string allocations during beatmap/storyboard parsing

### PR DESCRIPTION
**Pull request by EVAST9919**
Making a draft pr before spending any more time on it. It will be split into multiple prs down the line, so you can call this one just a reference/proof of concept/whatever.

`string.Split` allocates an array of new sub-strings which can be avoided by using `ReadonlySpan<char>.Split`. However this method acts a bit differently: it needs a `Span<Range>` to store indexes of resulting substrings within the source string, which we can allocate on a stack.
The main question is: is this an acceptable way to proceed?

Example of parsing world.execute(me) (Can be improved even further)
|master|pr|
|---|---|
|master|pr|

### CODE REVIEW

**Comment by bdach** [comment]
seems to basically be a duplicate of https://github.com/ppy/osu/pull/23282?

**Comment by peppy** [comment]
> seems to basically be a duplicate of #23282?

I was looking for this for at least 5 minutes and couldn't find it. Github search is pretty shoddy.

Either way, the direction is still correct and something we want. I'd cross-check with the other PR and likely bring this one up-to-date with any (better) changes from the other one.

Using the same benchmarks as the other PR would be hugely beneficial, as they showed marginal alloc reductions, which is not what's being claimed here (maybe due to the beatmap being benchmarked, though).

**Comment by bdach** [comment]
well @huoyaoyuan said in that other thread that he was going revisit, so tagging just to ensure no wasted effort on either side

**Comment by huoyaoyuan** [comment]
> well @huoyaoyuan said in that other thread that he was going to revisit, so tagging just to ensure no wasted effort on either side

Thanks. I was using the exact same approach (the new `Span.Split` method in .NET 8), with a little stylish differences. I lost some commits on my machine at home.

File: osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs [hunk 1]
```diff
@@ -456,52 +456,52 @@ private void handleEvent(string line)
...
+
         private void handleTimingPoint(string line)
         {
-            string[] split = line.Split(',');
+            var span = line.AsSpan();
+            Span<Range> ranges = stackalloc Range[span.GetSplitCount(comma)];
+            span.Split(ranges, comma);
 
-            double time = getOffsetTime(Parsing.ParseDouble(split[0].Trim()));
+            double time = getOffsetTime(Parsing.ParseDouble(span.Slice(ranges[0]).Trim()));
```
Comment (huoyaoyuan) [inline osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs:467]: Stylish preference here: `span.Slice(ranges[x])` or `span[ranges[x]]`. Two brackets for different meaning is somehow less readable.

File: osu.Game/Rulesets/Objects/Legacy/LegacyConverterExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+
+namespace osu.Game.Rulesets.Objects.Legacy
+{
+    public static class LegacyConverterExtensions
+    {
+        public static int GetSplitCount(this ReadOnlySpan<char> text, char separator)
```
Comment (huoyaoyuan) [inline osu.Game/Rulesets/Objects/Legacy/LegacyConverterExtensions.cs:10]: You can use `MemoryExtensions.Count` here. It's extra fast with full SIMD support for primitive types.

File: osu.Game/Rulesets/Objects/Legacy/LegacyConverterExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+                    count++;
+            }
+
+            return count + 1;
+        }
+
+        public static ReadOnlySpan<char> Slice(this ReadOnlySpan<char> text, Range range)
+            => text.Slice(range.Start.Value, range.Length());
+
+        public static int Length(this Range range) => range.End.Value - range.Start.Value;
```
Comment (huoyaoyuan) [inline osu.Game/Rulesets/Objects/Legacy/LegacyConverterExtensions.cs:26]: This is a bit unreliable. A range can contain index from the end.

You should just use the slice operator on span. If preferred, you can create an alias method for the operator (`Slice(span, range) => span[range]`) for readness.

File: osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs [hunk 4]
```diff
@@ -69,30 +75,30 @@ public override HitObject Parse(string text)
             {
                 result = CreateHit(pos, combo, comboOffset);
 
-                if (split.Length > 5)
-                    readCustomSampleBanks(split[5], bankInfo);
+                if (ranges.Length > 5)
```
Comment (huoyaoyuan) [inline osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs:78]: I was using an alternative approach: stackalloc the `Span<Range>` with maximum length we want to support (+1 for guarding), and check the return value of `Split`.
Both approaches should be functionally correct.

File: osu.Game/Beatmaps/Formats/LegacyStoryboardDecoder.cs [hunk 5]
```diff
@@ -167,150 +171,118 @@ private void handleEvents(string line)
                 if (depth < 2)
                     timelineGroup = storyboardSprite?.TimelineGroup;
 
-                string commandType = split[0];
+                var commandType = span.Slice(ranges[0]);
 
-                switch (commandType)
+                if (commandType.SequenceEqual("T"))
```
Comment (huoyaoyuan) [inline osu.Game/Beatmaps/Formats/LegacyStoryboardDecoder.cs:176]: You can unset the hard coded LangVersion=10 in osu.Game, and use the switch on span feature from C# 11.

**Comment by huoyaoyuan** [comment]
> Using the same benchmarks as the other PR would be hugely beneficial, as they showed marginal alloc reductions, which is not what's being claimed here (maybe due to the beatmap being benchmarked, though).

The majority of allocation should come from `TextReader.ReadLine`. It would allocate at least a string for each line.

To ultimately resolve that, the whole parsing infrastructure needs to be refactored. We can then parse on the raw UTF-8 `ReadOnlySpan<byte>` of the file. The `ReadOnlySpan<char>` based approach will be certainly reusable and won't be sunk effort.

**Comment by smoogipoo** [comment]
I feel like this is going in the wrong direction because it's keeping existing structures and working around it.

A few years ago I implemented my own attempt at this: https://github.com/smoogipoo/osu/tree/tokenizer-legacy-decoder
I think this matches a bit more to @huoyaoyuan 's implementation, and it's the path I want to see going forward - implementing a tokeniser for it. I think @huoyaoyuan 's was a bit more raw than mine if I remember correctly.

**Comment by EVAST9919** [comment]
I don't mind closing this in favour of https://github.com/ppy/osu/pull/23282
Also https://discord.com/channels/188630481301012481/188630652340404224/1211954046840213544

**Pull request closed without merging**